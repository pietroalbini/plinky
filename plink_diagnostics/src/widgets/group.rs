use super::WidgetWriter;
use crate::widgets::Widget;

pub struct WidgetGroup {
    widgets: Vec<Box<dyn Widget>>,
}

impl WidgetGroup {
    pub fn new() -> Self {
        WidgetGroup { widgets: Vec::new() }
    }

    pub fn add<T: Widget + 'static>(mut self, widget: T) -> Self {
        self.widgets.push(Box::new(widget));
        self
    }
}

impl Widget for WidgetGroup {
    fn render(&self, writer: &mut dyn WidgetWriter) {
        let mut indent_writer = IndentWriter {
            initial: true,
            last_char_is_newline: false,
            pattern: " │ ",
            inner: writer,
        };
        for (idx, widget) in self.widgets.iter().enumerate() {
            if idx == 0 {
                indent_writer.push_str("\n");
            } else {
                indent_writer.push_str("\n\n");
            }
            widget.render(&mut indent_writer);
            if indent_writer.last_char_is_newline {
                panic!("widgets must not terminate with a newline");
            }
        }
        if !self.widgets.is_empty() {
            writer.push_str("\n ┴");
        }
    }
}

struct IndentWriter<'a> {
    initial: bool,
    last_char_is_newline: bool,
    pattern: &'a str,
    inner: &'a mut dyn WidgetWriter,
}

impl WidgetWriter for IndentWriter<'_> {
    fn push(&mut self, content: char) {
        if self.initial {
            self.inner.push_str(self.pattern);
            self.initial = false;
        }
        self.inner.push(content);
        if content == '\n' {
            self.inner.push_str(self.pattern);
            self.last_char_is_newline = true;
        } else {
            self.last_char_is_newline = false;
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::configure_insta;
    use crate::widgets::{Table, Text};
    use insta::assert_snapshot;

    #[test]
    fn test_empty_group() {
        assert!(WidgetGroup::new().render_to_string().is_empty());
    }

    #[test]
    fn test_multiple_widgets() {
        let _config = configure_insta();

        let mut table = Table::new();
        table.add_row(["Foo", "Bar"]);

        let group = WidgetGroup::new().add(Text::new("A simple text message!")).add(table);
        assert_snapshot!(group.render_to_string());
    }
}
