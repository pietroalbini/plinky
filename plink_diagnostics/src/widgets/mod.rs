mod table;

pub use self::table::Table;
