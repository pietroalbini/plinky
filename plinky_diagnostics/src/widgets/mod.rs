mod group;
mod hex_dump;
mod quoted_text;
mod table;
mod text;

pub use self::group::WidgetGroup;
pub use self::hex_dump::HexDump;
pub use self::quoted_text::QuotedText;
pub use self::table::Table;
pub use self::text::Text;
use crate::WidgetWriter;

pub trait Widget {
    fn render(&self, writer: &mut WidgetWriter<'_>);

    fn render_to_string(&self) -> String {
        let mut buffer = String::new();
        let mut writer = WidgetWriter::new(&mut buffer);
        self.render(&mut writer);
        buffer
    }
}

impl Widget for Box<dyn Widget> {
    fn render(&self, writer: &mut WidgetWriter<'_>) {
        self.as_ref().render(writer)
    }
}
