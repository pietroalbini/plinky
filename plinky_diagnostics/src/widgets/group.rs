use crate::widgets::Widget;
use crate::WidgetWriter;
use crate::writer::IndentMode;

pub struct WidgetGroup {
    widgets: Vec<Box<dyn Widget>>,
}

impl WidgetGroup {
    pub fn new() -> Self {
        WidgetGroup { widgets: Vec::new() }
    }

    pub fn add<T: Widget + 'static>(mut self, widget: T) -> Self {
        self.widgets.push(Box::new(widget));
        self
    }
}

impl Widget for WidgetGroup {
    fn render(&self, writer: &mut WidgetWriter) {
        writer.push_indent(" │", IndentMode::ShowAlways);
        writer.push_indent("  ", IndentMode::HideOnEmptyLines);
        for (idx, widget) in self.widgets.iter().enumerate() {
            if idx == 0 {
                writer.push_str("\n");
            } else {
                writer.push_str("\n\n");
            }
            widget.render(writer);
            if writer.last_char() == Some('\n') {
                panic!("widgets must not terminate with a newline");
            }
        }
        writer.pop_indent();
        writer.pop_indent();
        if !self.widgets.is_empty() {
            writer.push_str("\n ┴");
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::configure_insta;
    use crate::widgets::{Table, Text};
    use insta::assert_snapshot;

    #[test]
    fn test_empty_group() {
        assert!(WidgetGroup::new().render_to_string().is_empty());
    }

    #[test]
    fn test_multiple_widgets() {
        let _config = configure_insta();

        let mut table = Table::new();
        table.add_row(["Foo", "Bar"]);

        let group = WidgetGroup::new().add(Text::new("A simple text message!")).add(table);
        assert_snapshot!(group.render_to_string());
    }
}
