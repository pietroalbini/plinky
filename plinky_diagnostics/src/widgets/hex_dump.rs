use crate::WidgetWriter;
use crate::widgets::{Table, Widget};
use std::fmt::Write;

const BYTES_PER_LINE: usize = 16;

pub struct HexDump {
    data: Vec<u8>,
}

impl HexDump {
    pub fn new(data: impl Into<Vec<u8>>) -> Self {
        Self { data: data.into() }
    }
}

impl Widget for HexDump {
    fn render(&self, writer: &mut WidgetWriter<'_>) {
        let mut hex = String::new();
        let mut ascii = String::new();

        for chunk in self.data.chunks(BYTES_PER_LINE) {
            for (i, byte) in chunk.iter().copied().enumerate() {
                if i > 0 {
                    hex.push(' ');
                }
                hex.write_fmt(format_args!("{byte:0>2x}")).unwrap();
                if show_as_ascii(byte) {
                    ascii.push(byte as char);
                } else {
                    ascii.push('.');
                }
            }
            hex.push('\n');
            ascii.push('\n');
        }

        let mut table = Table::new();
        table.add_row([hex.trim(), ascii.trim()]);
        table.render(writer);
    }
}

fn show_as_ascii(byte: u8) -> bool {
    byte.is_ascii_alphanumeric() || byte.is_ascii_punctuation() || byte == b' '
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::configure_insta;
    use insta::assert_snapshot;

    #[test]
    fn test_hex_dump_hello() {
        let _config = configure_insta();

        assert_snapshot!(HexDump::new(b"Hello world").render_to_string());
    }

    #[test]
    fn test_hex_dump_256() {
        let _config = configure_insta();

        let data = (0u8..=255u8).collect::<Vec<_>>();
        assert_snapshot!(HexDump::new(data).render_to_string());
    }
}
