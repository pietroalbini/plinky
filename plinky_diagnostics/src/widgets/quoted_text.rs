use crate::WidgetWriter;
use crate::widgets::Widget;
use crate::writer::IndentMode;

pub struct QuotedText {
    content: String,
}

impl QuotedText {
    pub fn new(content: impl Into<String>) -> Self {
        Self { content: content.into() }
    }
}

impl Widget for QuotedText {
    fn render(&self, writer: &mut WidgetWriter) {
        if self.content.is_empty() {
            return;
        }

        writer.push_str("╭\n");

        writer.push_indent("│", IndentMode::ShowAlways);
        writer.push_indent(" ", IndentMode::HideOnEmptyLines);

        writer.push_str(&self.content);
        writer.push('\n');

        writer.pop_indent();
        writer.pop_indent();

        writer.push_str("╰");
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::configure_insta;
    use insta::assert_snapshot;

    #[test]
    fn test_empty_content() {
        assert_eq!("", QuotedText::new("").render_to_string());
    }

    #[test]
    fn test_single_line() {
        let _config = configure_insta();

        let content = QuotedText::new("Hello world");
        assert_snapshot!(content.render_to_string());
    }

    #[test]
    fn test_multiple_lines() {
        let _config = configure_insta();

        let content = QuotedText::new("Hello world\nThis\n\n  has\n    multiple lines!");
        assert_snapshot!(content.render_to_string());
    }
}
