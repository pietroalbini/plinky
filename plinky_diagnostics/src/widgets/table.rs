use crate::widgets::Widget;
use crate::WidgetWriter;

const UNICODE_CHARSET: TableCharset = TableCharset {
    vertical_separator: '│',
    horizontal_separator: '─',
    first_junction: TableJunctionCharset { first: '╭', middle: '┬', last: '╮' },
    middle_junction: TableJunctionCharset { first: '├', middle: '┼', last: '┤' },
    last_junction: TableJunctionCharset { first: '╰', middle: '┴', last: '╯' },
};

pub struct Table {
    charset: &'static TableCharset,
    title: Option<String>,
    state: TableState,
}

impl Table {
    pub fn new() -> Self {
        Self { charset: &UNICODE_CHARSET, title: None, state: TableState::Empty }
    }

    pub fn set_title(&mut self, title: impl Into<String>) {
        self.title = Some(title.into());
    }

    pub fn add_row<I, V>(&mut self, row: I)
    where
        V: Into<String>,
        I: IntoIterator<Item = V>,
    {
        let mut count = 0;
        let row = row
            .into_iter()
            .map(|cell| cell.into().split('\n').map(|s| s.to_string()).collect::<Vec<_>>())
            .inspect(|_| count += 1);

        match &mut self.state {
            TableState::Empty => {
                let mut cells_len = Vec::new();
                self.state = TableState::HasContent {
                    content: row
                        .inspect(|cell| {
                            cells_len.push(cell.iter().map(|l| l.len()).max().unwrap_or(0))
                        })
                        .collect(),
                    cells_len,
                    cells_count: count,
                };
            }
            TableState::HasContent { cells_count, cells_len, content } => {
                content.extend(row.enumerate().map(|(pos, cell)| {
                    let cell_len = cell.iter().map(|c| c.len()).max().unwrap_or(0);
                    match cells_len.get_mut(pos) {
                        Some(len) => *len = (*len).max(cell_len),
                        None => cells_len.push(cell_len),
                    }
                    cell
                }));
                if count != *cells_count {
                    panic!("other rows have {cells_count} cells, while this row has {count} cells");
                }
            }
        }
    }

    fn render_horizontal_border(
        &self,
        writer: &mut WidgetWriter,
        cells_len: &[usize],
        junction: &TableJunctionCharset,
    ) {
        writer.push(junction.first);
        for (idx, len) in cells_len.iter().enumerate() {
            if idx != 0 {
                writer.push(junction.middle);
            }
            for _ in 0..(*len + 2) {
                writer.push(self.charset.horizontal_separator);
            }
        }
        writer.push(junction.last);
    }
}

impl Widget for Table {
    fn render(&self, writer: &mut WidgetWriter) {
        let TableState::HasContent { cells_count, cells_len, content } = &self.state else {
            panic!("trying to render an empty table");
        };

        if let Some(title) = &self.title {
            writer.push_str("  ");
            writer.push_str(title);
            writer.push_str("\n");
        }
        self.render_horizontal_border(writer, cells_len, &self.charset.first_junction);
        writer.push_str("\n");

        let mut rows = content.chunks_exact(*cells_count);
        let mut idx = 0;
        for row in rows.by_ref() {
            let last_row = idx == content.len() / cells_count - 1;
            let lines_count = row.iter().map(|cell| cell.len()).max().unwrap_or(0);

            for line in 0..lines_count {
                writer.push(self.charset.vertical_separator);
                for (idx, cell_all_lines) in row.iter().enumerate() {
                    let cell = cell_all_lines.get(line).map(|c| c.as_str()).unwrap_or_default();
                    writer.push(' ');
                    writer.push_str(cell);

                    // Padding to align all cells.
                    for _ in cell.len()..cells_len[idx] {
                        writer.push(' ');
                    }

                    writer.push(' ');
                    writer.push(self.charset.vertical_separator);
                }
                writer.push('\n');
            }
            self.render_horizontal_border(
                writer,
                cells_len,
                if last_row { &self.charset.last_junction } else { &self.charset.middle_junction },
            );
            if !last_row {
                writer.push('\n');
            }

            idx += 1;
        }

        assert!(rows.remainder().is_empty());
    }
}

enum TableState {
    Empty,
    HasContent { cells_count: usize, cells_len: Vec<usize>, content: Vec<Vec<String>> },
}

struct TableCharset {
    vertical_separator: char,
    horizontal_separator: char,
    first_junction: TableJunctionCharset,
    middle_junction: TableJunctionCharset,
    last_junction: TableJunctionCharset,
}

struct TableJunctionCharset {
    first: char,
    middle: char,
    last: char,
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::configure_insta;
    use insta::assert_snapshot;

    #[test]
    fn test_sample_table() {
        let _config = configure_insta();

        let mut table = Table::new();
        table.add_row(["Foo", "Bar", "Baz"]);
        table.add_row(["Hello", "super long", "world!"]);
        table.add_row(["98%", "", "-"]);

        assert_snapshot!(table.render_to_string());
    }

    #[test]
    fn test_single_cell_table() {
        let _config = configure_insta();

        let mut table = Table::new();
        table.add_row(["alone"]);

        assert_snapshot!(table.render_to_string());
    }

    #[test]
    fn test_table_with_title() {
        let _config = configure_insta();

        let mut table = Table::new();
        table.set_title("Example title:");
        table.add_row(["a", "b", "c"]);

        assert_snapshot!(table.render_to_string());
    }

    #[test]
    fn test_table_with_multiple_lines() {
        let _config = configure_insta();

        let mut table = Table::new();
        table.add_row(["a", "b", "c"]);
        table.add_row(["foo\nbar", "baz", "qu\nu\n\n\nx!!!!!!!!"]);

        assert_snapshot!(table.render_to_string());
    }

    #[test]
    #[should_panic = "other rows have 3 cells, while this row has 2 cells"]
    fn test_wrong_number_of_cells() {
        let mut table = Table::new();
        table.add_row(["foo", "bar", "baz"]);
        table.add_row(["a", "b"]);
    }

    #[test]
    #[should_panic = "trying to render an empty table"]
    fn test_empty_table() {
        Table::new().render_to_string();
    }
}
