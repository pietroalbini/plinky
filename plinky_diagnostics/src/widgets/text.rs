use crate::widgets::Widget;
use crate::WidgetWriter;

pub struct Text {
    content: String,
}

impl Text {
    pub fn new(content: impl Into<String>) -> Self {
        Self { content: content.into() }
    }
}

impl Widget for Text {
    fn render(&self, writer: &mut WidgetWriter) {
        writer.push_str(&self.content);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_render_text() {
        assert_eq!("Hello world!", Text::new("Hello world!").render_to_string());
    }
}
