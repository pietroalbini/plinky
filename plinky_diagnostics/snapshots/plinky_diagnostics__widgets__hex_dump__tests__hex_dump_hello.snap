---
source: plinky_diagnostics/src/widgets/hex_dump.rs
expression: "HexDump::new(b\"Hello world\").render_to_string()"
---
╭──────────────────────────────────┬─────────────╮
│ 48 65 6c 6c 6f 20 77 6f 72 6c 64 │ Hello world │
╰──────────────────────────────────┴─────────────╯
